Add RetryTestRunner

rogozinds opened the pull request:
RetryTestRunner allows to run tests several times.
If tests passes at least once it is mared as passed.
That can be useful to prevent random failures.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/testbench/852)
<!-- Reviewable:end -->

Artur- commented:
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYXuAKol78UkK4kG_0f:-KYXuAKol78UkK4kG_0g:b-otbamo) (no related file):*
As discussed, this should be done as a `@Rule` instead so you can use it together with any runner

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/testbench/852)*
<!-- Sent from Reviewable.io -->

alvarezguille commented:
*[a discussion](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYXuAKol78UkK4kG_0f:-KYYddivdeHCy3NuXR17:b-31k649) (no related file):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

As discussed, this should be done as a `@Rule` instead so you can use it together with any runner
</blockquote></details>

What's the use case we want to cover? Something general for all tests? or particular case for counted tests?

Should this be a rule by default in TestBenchTestCase with 0 retries by default and the possibility to use a system property to set retries?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/testbench/852)*
<!-- Sent from Reviewable.io -->

rogozinds commented:
Review status: 0 of 11 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYXuAKol78UkK4kG_0f:-KYlx6Op-T9uo0SmFqtM:bgend5c) (no related file):*
<details><summary><i>Previously, alvarezguille (Guille) wrote…</i></summary><blockquote>

What's the use case we want to cover? Something general for all tests? or particular case for counted tests?

Should this be a rule by default in TestBenchTestCase with 0 retries by default and the possibility to use a system property to set retries?
</blockquote></details>

Changed to rule

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/testbench/852)*
<!-- Sent from Reviewable.io -->

Artur- commented:
Reviewed 1 of 6 files at r2, 6 of 10 files at r3, 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 21 unresolved discussions.

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/Parameters.java, line 430 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlxC_oQb8hJ4MDOshj:-KYlxC_oQb8hJ4MDOshk:b-o7xqa1) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/main/java/com/vaadin/testbench/Parameters.java#L430)):*
> ```Java
>                count = Integer.parseInt(countParam);
>             } catch (NumberFormatException e) {
>                // NOP count == 1 by default
> ```

Tell the user he has a typo and allow him to fix it instead of keeping quiet and pretending all is well

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/TestBenchTestCase.java, line 43 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlyhggVtnCTJ-cEye3:-KYlyhggVtnCTJ-cEye4:b-70o4kd) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/main/java/com/vaadin/testbench/TestBenchTestCase.java#L43)):*
> ```Java
>     /**
>      * Specifies retry count, which is used to run same test several times.
>      * Can be changed by setting retrycount system property.
> ```

the "insert.correct.property.name.here" system property

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/TestBenchTestCase.java, line 44 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlzB4otA8aEjOnS3fM:-KYlzB4otA8aEjOnS3fN:bvjl76m) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/main/java/com/vaadin/testbench/TestBenchTestCase.java#L44)):*
> ```Java
>      * Specifies retry count, which is used to run same test several times.
>      * Can be changed by setting retrycount system property.
>      * Can be overriden in a child class.
> ```

How can it be overridden?

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/TestBenchTestCase.java, line 45 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlzSdOx0TcTL-1AugG:-KYlzSdOx0TcTL-1AugH:bsdxy4o) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/main/java/com/vaadin/testbench/TestBenchTestCase.java#L45)):*
> ```Java
>      * Can be changed by setting retrycount system property.
>      * Can be overriden in a child class.
>      * Default: 1
> ```

Why is a test retried one time per default? It should be zero

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 7 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlz2M9PCX3PbVsA3DM:-KYlz2M9PCX3PbVsA3DN:bqkw7qm) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L7)):*
> ```Java
> import org.junit.runners.model.Statement;
> 
> public class RetryRule implements TestRule {
> ```

javadoc?

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 10 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlz3hk_9UlhTmGdNwy:-KYlz3hk_9UlhTmGdNwz:bqkw7qm) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L10)):*
> ```Java
>     private int retryCount;
> 
>     public RetryRule(int retryCount) {
> ```

javadoc?

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 14 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlz4OF5_y_fJsfxLMm:-KYlz4OF5_y_fJsfxLMn:bqkw7qm) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L14)):*
> ```Java
>     }
> 
>     public Statement apply(Statement base, Description description) {
> ```

javadoc?

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 29 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlz_qpb-HVEXBlvzhE:-KYlz_qpb-HVEXBlvzhF:bs6bfs3) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L29)):*
> ```Java
>                 Throwable caughtThrowable = null;
>                 // implement retry logic here
>                 for (int i = 0; i < retryCount; i++) {
> ```

Looks like the statement is never evaluated if you disable retries, i.e. retryCount==0?

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 37 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlzJ37_qzLJWEyf5wj:-KYlzJ37_qzLJWEyf5wk:bo9n6y7) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L37)):*
> ```Java
>                     }
>                 }
>                 System.err.println(String.format("%s: run failed %s times",
> ```

Do not use `System.err.println`

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 44 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlzg42NJ3GJR5vjnS2:-KYlzg42NJ3GJR5vjnS3:bbtxm0z) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L44)):*
> ```Java
>         };
>     }
> }
> ```

Add newline at end of file

---

*[vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/ParentRetryRuleTest.java, line 19 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlzsAo297MACZ6uiaN:-KYlzsAo297MACZ6uiaO:b-6865gq) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/ParentRetryRuleTest.java#L19)):*
> ```Java
>     public void parentRule_settingRule_overridesParentRule() {
>         passes = !passes;
>         Assert.assertTrue(passes);
> ```

What does this assert/test?

---

*[vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleAfterTest.java, line 22 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYm-3WIpb554k-rwJNS:-KYm-3WIpb554k-rwJNT:bnjkadp) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleAfterTest.java#L22)):*
> ```Java
>     @Test
>     public void afterAnnotation_executedSeveralTimes_whenUsingRetryRule() {
>         Assert.assertEquals(2, afterExecutedCount);
> ```

What does this test? That cleanUp was executed at least 2 times when retryCount is 3?

---

*[vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleBeforeTest.java, line 9 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYm-Dg_l1pREtIgIuqh:-KYm-Dg_l1pREtIgIuqi:b-8d3go2) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleBeforeTest.java#L9)):*
> ```Java
> 
> 
> public class RetryRuleBeforeTest {
> ```

This could probably be the same test class as `RetryRuleAfterTest`? 

---

*[vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleBeforeTest.java, line 22 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYm-BgdJw9_Ha__UBGB:-KYm-BgdJw9_Ha__UBGC:b191yyh) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleBeforeTest.java#L22)):*
> ```Java
>     @Test
>     public void beforeAnnotation_executedSeveralTimes_whenUsingRetryRule() {
>         Assert.assertEquals(2, beforeExecutedCount);
> ```

What does this test? That setUp was executed at least 2 times when retryCount is 2? 

---

*[vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleTest.java, line 17 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYm-TocdJC_R1COfXv9:-KYm-TodjajC1ojXBXmD:bk3e8t9) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleTest.java#L17)):*
> ```Java
>     @Test
>     public void randomFailureWithRetry_passesOnce_testPasses() {
>         int rand = new Random().nextInt(9);
> ```

Don't use random in tests, it only makes it hard to debug. Also you have no guarantee that this test actually passes, `rand` can e.g. be `1` 10 times in a row.

---

*[vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleTest.java, line 23 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYm-hpD4_1o2fEWP5cZ:-KYm-hpD4_1o2fEWP5c_:b263whn) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleTest.java#L23)):*
> ```Java
>     @Test(expected = AssertionError.class)
>     public void randomFailureWithRetry_notPassesOnce_testFails() {
>         Assert.assertTrue(false);
> ```

What does this test?

---

*[vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleTestbenchTest.java, line 9 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYm0xmhltFHTMVxbSo3:-KYm-mC9NHTAGlDQTWPk:b-11ksi5) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleTestbenchTest.java#L9)):*
> ```Java
> 
> 
> public class RetryRuleTestbenchTest extends TestBenchTestCase {
> ```

Does this test provide any additional info compared to `RetryRuleTest`?

---

*[vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleTestbenchWIthParameterTest.java, line 14 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYm0UKgf4r6sXGrON2Q:-KYm0UKgf4r6sXGrON2R:bx6dqxy) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleTestbenchWIthParameterTest.java#L14)):*
> ```Java
> 
>     static {
>         System.setProperty(Parameters.class.getName() + "." + "retryCount", "2");
> ```

This affects all other tests which are run in other threads and will probably cause random failures

---

*[vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleTestbenchWIthParameterTest.java, line 29 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYm0ePXptc7o_DPpuve:-KYm0ePXptc7o_DPpuvf:b-eurkta) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleTestbenchWIthParameterTest.java#L29)):*
> ```Java
>     public void cleanUp() {
>         System.getProperties().remove(Parameters.class.getName() + "." + "retryCount");
>         System.getProperty(Parameters.class.getName() + "." + "retryCount");
> ```

What does this do?

---

*[vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/WithoutRetryTest.java, line 10 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYm0jgqcYuhryQBX7Ch:-KYm0jgqcYuhryQBX7Ci:bceduzh) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/WithoutRetryTest.java#L10)):*
> ```Java
> 
>     @Test(expected = AssertionError.class)
>     public void defaultExecution_noRetryRule_testcaseRunOnce() {
> ```

What does this test except that JUnit works?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/testbench/852)*
<!-- Sent from Reviewable.io -->

rogozinds commented:
Review status: all files reviewed at latest revision, 21 unresolved discussions.

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/Parameters.java, line 430 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlxC_oQb8hJ4MDOshj:-KYrHIdc3c8F6K4M6WdT:b-8qwmq3) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/main/java/com/vaadin/testbench/Parameters.java#L430)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

Tell the user he has a typo and allow him to fix it instead of keeping quiet and pretending all is well
</blockquote></details>

Throw a RuntimeException

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/TestBenchTestCase.java, line 43 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlyhggVtnCTJ-cEye3:-KYrUsFGgRgNK3x7C11_:b-896fix) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/main/java/com/vaadin/testbench/TestBenchTestCase.java#L43)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

the "insert.correct.property.name.here" system property
</blockquote></details>

Done.

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/TestBenchTestCase.java, line 44 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlzB4otA8aEjOnS3fM:-KYrUr3vI6h0fsBtm_oe:b-n9tkhd) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/main/java/com/vaadin/testbench/TestBenchTestCase.java#L44)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

How can it be overridden?
</blockquote></details>

changed

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/TestBenchTestCase.java, line 45 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlzSdOx0TcTL-1AugG:-KYrUtWuvFVm5SBok7x6:bilg66l) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/main/java/com/vaadin/testbench/TestBenchTestCase.java#L45)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

Why is a test retried one time per default? It should be zero
</blockquote></details>

Now should be clear as it is max.attempts

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 7 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlz2M9PCX3PbVsA3DM:-KYrN2cq1EJHDIztWfFa:b-8b11j9) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L7)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

javadoc?
</blockquote></details>

Added

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 10 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlz3hk_9UlhTmGdNwy:-KYrVC74sSv9ZTwr83rt:b-7tfmjp) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L10)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

javadoc?
</blockquote></details>

added

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 14 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlz4OF5_y_fJsfxLMm:-KYrVG5JI0WUhd8NY8yM:b-qws8cj) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L14)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

javadoc?
</blockquote></details>

overriden

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 29 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlz_qpb-HVEXBlvzhE:-KYrOla__JfilnR1CMPj:b-896fix) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L29)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

Looks like the statement is never evaluated if you disable retries, i.e. retryCount==0?
</blockquote></details>

Done.

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 37 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlzJ37_qzLJWEyf5wj:-KYrVeYaK3U55QQtKmKo:b5xx1qc) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L37)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

Do not use `System.err.println`
</blockquote></details>

Changed to throw an exception.

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 44 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlzg42NJ3GJR5vjnS2:-KYrN0oCum5xbxKjOwtJ:b-896fix) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L44)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

Add newline at end of file
</blockquote></details>

Done.

---

*[vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/ParentRetryRuleTest.java, line 19 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYlzsAo297MACZ6uiaN:-KYrKaFS0eKDMYlmkw0S:beu7l2k) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/ParentRetryRuleTest.java#L19)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

What does this assert/test?
</blockquote></details>

Changed to use int.

---

*[vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleAfterTest.java, line 22 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYm-3WIpb554k-rwJNS:-KYrKlewLyKNB_nu92YH:b-rzg87l) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleAfterTest.java#L22)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

What does this test? That cleanUp was executed at least 2 times when retryCount is 3?
</blockquote></details>

It tests that method with @After annotation is executed on every retry

---

*[vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleBeforeTest.java, line 9 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYm-Dg_l1pREtIgIuqh:-KYrKtWWyjJdLC0FNYIw:b-ti3hon) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleBeforeTest.java#L9)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

This could probably be the same test class as `RetryRuleAfterTest`? 
</blockquote></details>

No it can't because it uses a static variable. And it can have only one test, otherwise method with @after
annotation would be executed retryCount*numberOfTests.

---

*[vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleBeforeTest.java, line 22 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYm-BgdJw9_Ha__UBGB:-KYrLCFXk3LHknko3oNN:bd9omp4) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleBeforeTest.java#L22)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

What does this test? That setUp was executed at least 2 times when retryCount is 2? 
</blockquote></details>

It tests that method with @Before annotation is executed on every retry

---

*[vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleTest.java, line 17 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYm-TocdJC_R1COfXv9:-KYrWLeuusVscb7AbL5y:ba3dfob) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleTest.java#L17)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

Don't use random in tests, it only makes it hard to debug. Also you have no guarantee that this test actually passes, `rand` can e.g. be `1` 10 times in a row.
</blockquote></details>

Removed

---

*[vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleTest.java, line 23 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYm-hpD4_1o2fEWP5cZ:-KYrWKdpk2lDtIthjn_0:ba3dfob) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleTest.java#L23)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

What does this test?
</blockquote></details>

Removed

---

*[vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleTestbenchTest.java, line 9 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYm0xmhltFHTMVxbSo3:-KYrWe9uYZ89DUfMWZ63:b-aoq84r) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleTestbenchTest.java#L9)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

Does this test provide any additional info compared to `RetryRuleTest`?
</blockquote></details>

Tests that default behaviour for TestbenchTestCase is running once. Changed to use int.

---

*[vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleTestbenchWIthParameterTest.java, line 14 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYm0UKgf4r6sXGrON2Q:-KYr_eG8S0xUN1JMpwNU:bbam269) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleTestbenchWIthParameterTest.java#L14)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

This affects all other tests which are run in other threads and will probably cause random failures
</blockquote></details>

That's not gonna be a problem, as it won't run in parallel. Because it's not extending the ParallelTest.
It should be inside static block, that it will be setup before class init.

---

*[vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleTestbenchWIthParameterTest.java, line 29 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYm0ePXptc7o_DPpuve:-KYr_wpWVC0EYyuNq49-:b-2x8sg7) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/RetryRuleTestbenchWIthParameterTest.java#L29)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

What does this do?
</blockquote></details>

Removed last line

---

*[vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/WithoutRetryTest.java, line 10 at r4](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYm0jgqcYuhryQBX7Ch:-KYrLz7P4VhKpN62UVn-:b-hfpv0l) ([raw file](https://github.com/vaadin/testbench/blob/2c63eb72d84e15de2ee5b930477ed987e8c61514/vaadin-testbench-core/src/test/java/com/vaadin/testbench/annotations/WithoutRetryTest.java#L10)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

What does this test except that JUnit works?
</blockquote></details>

removed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/testbench/852)*
<!-- Sent from Reviewable.io -->

Artur- commented:
Reviewed 1 of 6 files at r2, 11 of 11 files at r5.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/Parameters.java, line 421 at r5](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYu7f73BagXVvHxBPj8:-KYu7f741jg1S8vsDdnt:b-wc7ktj) ([raw file](https://github.com/vaadin/testbench/blob/9325d66d62cec872d2f2fdba0b49868bc4b16ecf/vaadin-testbench-core/src/main/java/com/vaadin/testbench/Parameters.java#L421)):*
> ```Java
>     /**
>      *
>      * @return maximum attempts the test can be run.
> ```

Include a summary sentence

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 8 at r5](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYu5413liDAyNB0RSFm:-KYu5413liDAyNB0RSFn:be32nzb) ([raw file](https://github.com/vaadin/testbench/blob/9325d66d62cec872d2f2fdba0b49868bc4b16ecf/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L8)):*
> ```Java
> 
> /**
>  * A RetryRule specifies retry count, that is used to run same test several times.
> ```

"A retry rule is used to re-run a test several times in case of a random failure. The test passes as soon as one attempt is executed without any errors, i.e. it is only run as many times as needed. The maximum number of attempts is specified in the constructor."

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 18 at r5](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYu5wj4qaG6UUaGAEAf:-KYu5wj4qaG6UUaGAEAg:b-b0aqu6) ([raw file](https://github.com/vaadin/testbench/blob/9325d66d62cec872d2f2fdba0b49868bc4b16ecf/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L18)):*
> ```Java
>     private int maxAttempts;
> 
>     /**
> ```

Leaving out the summary sentence will make all generated javadocs look ugly

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 20 at r5](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYu5cplyOE-YEstEIWA:-KYu5cpmsUYM5yUIAz95:bb8b5th) ([raw file](https://github.com/vaadin/testbench/blob/9325d66d62cec872d2f2fdba0b49868bc4b16ecf/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L20)):*
> ```Java
>     /**
>      *
>      * @param maxAttempts specifies the amount of maximum attempts for running test
> ```

"[...] the maximum number of times to run the test"

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 28 at r5](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYu5gxwFxvqaVfRD0GN:-KYu5gxwFxvqaVfRD0GO:b-o7m9jx) ([raw file](https://github.com/vaadin/testbench/blob/9325d66d62cec872d2f2fdba0b49868bc4b16ecf/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L28)):*
> ```Java
>     /**
>      *
>      * @return the amount of maximum attempts for running test
> ```

Gets the maximum number of times to run the test.

`@return` the maximum number of times to run the test

Leaving out the summary sentence will make all generated javadocs look ugly

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/testbench/852)*
<!-- Sent from Reviewable.io -->

rogozinds commented:
Review status: 8 of 9 files reviewed at latest revision, 6 unresolved discussions.

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/Parameters.java, line 421 at r5](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYu7f73BagXVvHxBPj8:-KYwPxCLVST_O54_wNZV:b-896fix) ([raw file](https://github.com/vaadin/testbench/blob/9325d66d62cec872d2f2fdba0b49868bc4b16ecf/vaadin-testbench-core/src/main/java/com/vaadin/testbench/Parameters.java#L421)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

Include a summary sentence
</blockquote></details>

Done.

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 8 at r5](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYu5413liDAyNB0RSFm:-KYwISI9X4Q7_ctbr1ar:b-896fix) ([raw file](https://github.com/vaadin/testbench/blob/9325d66d62cec872d2f2fdba0b49868bc4b16ecf/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L8)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

"A retry rule is used to re-run a test several times in case of a random failure. The test passes as soon as one attempt is executed without any errors, i.e. it is only run as many times as needed. The maximum number of attempts is specified in the constructor."
</blockquote></details>

Done.

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 18 at r5](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYu5wj4qaG6UUaGAEAf:-KYwQ2so2V1HgLTBqgfq:b-896fix) ([raw file](https://github.com/vaadin/testbench/blob/9325d66d62cec872d2f2fdba0b49868bc4b16ecf/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L18)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

Leaving out the summary sentence will make all generated javadocs look ugly
</blockquote></details>

Done.

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 20 at r5](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYu5cplyOE-YEstEIWA:-KYwS4Y_ydoeTP0V1xdm:b-896fix) ([raw file](https://github.com/vaadin/testbench/blob/9325d66d62cec872d2f2fdba0b49868bc4b16ecf/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L20)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

"[...] the maximum number of times to run the test"
</blockquote></details>

Done.

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 28 at r5](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYu5gxwFxvqaVfRD0GN:-KYwS-fJ44VR0JXaQvx8:b-896fix) ([raw file](https://github.com/vaadin/testbench/blob/9325d66d62cec872d2f2fdba0b49868bc4b16ecf/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L28)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

Gets the maximum number of times to run the test.

`@return` the maximum number of times to run the test

Leaving out the summary sentence will make all generated javadocs look ugly
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/testbench/852)*
<!-- Sent from Reviewable.io -->

Artur- commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/Parameters.java, line 421 at r5](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYu7f73BagXVvHxBPj8:-KYwTELxzkJppIX42gu7:b-irwhab) ([raw file](https://github.com/vaadin/testbench/blob/9325d66d62cec872d2f2fdba0b49868bc4b16ecf/vaadin-testbench-core/src/main/java/com/vaadin/testbench/Parameters.java#L421)):*
<details><summary><i>Previously, rogozinds (rogozinds) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

where?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/testbench/852)*
<!-- Sent from Reviewable.io -->

rogozinds commented:
Review status: 7 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/Parameters.java, line 421 at r5](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYu7f73BagXVvHxBPj8:-KYx0XXDJQLc4IMx3JFc:br2bu3q) ([raw file](https://github.com/vaadin/testbench/blob/9325d66d62cec872d2f2fdba0b49868bc4b16ecf/vaadin-testbench-core/src/main/java/com/vaadin/testbench/Parameters.java#L421)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

where?
</blockquote></details>

Now it's added

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/testbench/852)*
<!-- Sent from Reviewable.io -->

alvarezguille commented:
Review status: 7 of 10 files reviewed at latest revision, 3 unresolved discussions.

---

*[documentation/testbench-known-issues.asciidoc, line 48 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYx7lFHXmVKlb2hhyEf:-KYx7lFHXmVKlb2hhyEg:b-ym1as8) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/documentation/testbench-known-issues.asciidoc#L48)):*
> ```AsciiDoc
> Internet Explorer specification requires browser window to be focused to receive user events,
> while Selenium Web Driver specification points out that Web Driver should work without window focus.
> ```

Introducing the issue in known issues looks ok. But I think the rest of the documentation for rule and system property should go somewhere else and be linked from here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/testbench/852)*
<!-- Sent from Reviewable.io -->

alvarezguille commented:
Reviewed 1 of 3 files at r7.
Review status: 8 of 10 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/testbench/852)*
<!-- Sent from Reviewable.io -->

Artur- commented:
Reviewed 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[documentation/testbench-known-issues.asciidoc, line 44 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxS8Qjq4e8iMaXBKa5:-KYxS8Qk7BLvf9nCBFIt:bgnx99y) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/documentation/testbench-known-issues.asciidoc#L44)):*
> ```AsciiDoc
> == Random Failures with Internet Explorer Driver
> 
> Unfortunately, sometimes TestBench test may behave unexpectedly, causing random failures.
> ```

, a TestBench test may sometimes behave unexpectedly

---

*[documentation/testbench-known-issues.asciidoc, line 47 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxSUHsEmGep0HZC-Uy:-KYxSUHsEmGep0HZC-Uz:b-y3ikyt) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/documentation/testbench-known-issues.asciidoc#L47)):*
> ```AsciiDoc
> One of the reasons for these random failures is problems with the Internet Explorer Driver.
> Internet Explorer specification requires browser window to be focused to receive user events,
> while Selenium Web Driver specification points out that Web Driver should work without window focus.
> ```

Could add `https://github.com/SeleniumHQ/selenium/wiki/InternetExplorerDriver#native-events-and-internet-explorer`

---

*[documentation/testbench-known-issues.asciidoc, line 50 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxSg-Rbrx3yVngInaW:-KYxSg-Rbrx3yVngInaX:b-s83tn6) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/documentation/testbench-known-issues.asciidoc#L50)):*
> ```AsciiDoc
> 
> TestBench provides a workaround to this problem.
> Use [classname]#RetryRule(int maxAttempts)# rule to re-run a test several times in case of a random failure.
> ```

If `TestBenchTestCase` already contains the `@Rule`, shouldn't this rather say that you can modify the number of attempts using the system property as the primary way? 

---

*[documentation/testbench-known-issues.asciidoc, line 65 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxSuSJCTjhimeM7N6x:-KYxSuSK47D3K-u4VRFb:ba6euwj) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/documentation/testbench-known-issues.asciidoc#L65)):*
> ```AsciiDoc
>     public void defaultExecution_noRetryRule_testcaseRunOnce() {
>         count++;
>         Assert.assertEquals(2, count);
> ```

This example is hard to understand - I would just show the `@Rule` part with a comment e.g. `// Run the test two times`

---

*[documentation/testbench-known-issues.asciidoc, line 73 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxT8r-Gld9R5TUeh2Y:-KYxT8r-Gld9R5TUeh2Z:byc2h34) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/documentation/testbench-known-issues.asciidoc#L73)):*
> ```AsciiDoc
> [classname]#RetryRule# affects all the test methods in the class and also child classes.
> 
> The default value of [parameter]#maxAttempts# is 1, meaning that tests run only once by default.
> ```

remove the second "by default".

---

*[documentation/testbench-known-issues.asciidoc, line 74 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxTCymaGdaurycFWXJ:-KYxTCymaGdaurycFWXK:b-tgz750) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/documentation/testbench-known-issues.asciidoc#L74)):*
> ```AsciiDoc
> 
> The default value of [parameter]#maxAttempts# is 1, meaning that tests run only once by default.
> You can change the value of [parameter]#maxAttempts# using Java system property:
> ```

the Java system property

---

*[documentation/testbench-known-issues.asciidoc, line 75 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxTHlGH7lUtVXxPV74:-KYxTHlGH7lUtVXxPV75:b5u7ryx) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/documentation/testbench-known-issues.asciidoc#L75)):*
> ```AsciiDoc
> The default value of [parameter]#maxAttempts# is 1, meaning that tests run only once by default.
> You can change the value of [parameter]#maxAttempts# using Java system property:
> [methodname]#System.setProperty(3, "max.attempts"))#. This will affect all the TestBench tests.
> ```

`-Dcom.vaadin.testbench.Parameters.maxAttempts=2` is the primary way. I cannot see why somebody would want to use `System.setProperty` for this, especially if you add `Parameters.setMaxAttempts(2)`.

---

*[documentation/testbench-known-issues.asciidoc, line 78 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxU1xYBH8WY0CLHu8b:-KYxU1xYBH8WY0CLHu8c:b-5en73z) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/documentation/testbench-known-issues.asciidoc#L78)):*
> ```AsciiDoc
> 
> [NOTE]
> Use [classname]#RetryRule# when you are sure that the test fails, because of the problems
> ```

remove ","

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/Parameters.java, line 426 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxTswP7qKU-NaMGxCI:-KYxTswP7qKU-NaMGxCJ:b5yoktb) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/vaadin-testbench-core/src/main/java/com/vaadin/testbench/Parameters.java#L426)):*
> ```Java
>      * @return maximum attempts the test can be run.
>      */
>     public static int getMaxAttempts() {
> ```

All other parameters also have a setter

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/Parameters.java, line 427 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxTc6nh-92sD0MlLUw:-KYxTc6nh-92sD0MlLUx:bmhb79c) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/vaadin-testbench-core/src/main/java/com/vaadin/testbench/Parameters.java#L427)):*
> ```Java
>      */
>     public static int getMaxAttempts() {
>         String countParam = System.getProperty(getQualifiedParameter("max.attempts"));
> ```

this should be maxAttempts to be consistent with all other parameters

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/testbench/852)*
<!-- Sent from Reviewable.io -->

rogozinds commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[documentation/testbench-known-issues.asciidoc, line 44 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxS8Qjq4e8iMaXBKa5:-KYx_8iT-o18lwNlu6ri:b-896fix) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/documentation/testbench-known-issues.asciidoc#L44)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

, a TestBench test may sometimes behave unexpectedly
</blockquote></details>

Done.

---

*[documentation/testbench-known-issues.asciidoc, line 47 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxSUHsEmGep0HZC-Uy:-KYxa6sBTkRO5dBDDwQq:bb6un1j) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/documentation/testbench-known-issues.asciidoc#L47)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

Could add `https://github.com/SeleniumHQ/selenium/wiki/InternetExplorerDriver#native-events-and-internet-explorer`
</blockquote></details>

Done

---

*[documentation/testbench-known-issues.asciidoc, line 48 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYx7lFHXmVKlb2hhyEf:-KYxaFCHZAFrwiYMX22U:b-qf0s7v) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/documentation/testbench-known-issues.asciidoc#L48)):*
<details><summary><i>Previously, alvarezguille (Guille) wrote…</i></summary><blockquote>

Introducing the issue in known issues looks ok. But I think the rest of the documentation for rule and system property should go somewhere else and be linked from here
</blockquote></details>

Suggestions where to put it? We were thinking to put it into best practices, but it's not actually the best practise ;)

---

*[documentation/testbench-known-issues.asciidoc, line 50 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxSg-Rbrx3yVngInaW:-KYxaR0dzWjFWZOHX5mA:b-6uyt3h) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/documentation/testbench-known-issues.asciidoc#L50)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

If `TestBenchTestCase` already contains the `@Rule`, shouldn't this rather say that you can modify the number of attempts using the system property as the primary way? 
</blockquote></details>

You don't want to do that, because it will be applied to all your tests. That's probably really rare use case (like with FW), and I personally don't like the idea of adding this rule to all your tests, it complete removes testing of random failures, that you might actually have in your application.

---

*[documentation/testbench-known-issues.asciidoc, line 65 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxSuSJCTjhimeM7N6x:-KYxat8bSQ4ieSRmB-Th:bb6un1j) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/documentation/testbench-known-issues.asciidoc#L65)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

This example is hard to understand - I would just show the `@Rule` part with a comment e.g. `// Run the test two times`
</blockquote></details>

Done

---

*[documentation/testbench-known-issues.asciidoc, line 73 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxT8r-Gld9R5TUeh2Y:-KYxbD0PuSldJ1zdUH2D:b-896fix) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/documentation/testbench-known-issues.asciidoc#L73)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

remove the second "by default".
</blockquote></details>

Done.

---

*[documentation/testbench-known-issues.asciidoc, line 74 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxTCymaGdaurycFWXJ:-KYxbNtmA1lwF-UO75dz:b-896fix) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/documentation/testbench-known-issues.asciidoc#L74)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

the Java system property
</blockquote></details>

Done.

---

*[documentation/testbench-known-issues.asciidoc, line 75 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxTHlGH7lUtVXxPV74:-KYxcmRWale2QqpAKj13:b-jjsbpw) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/documentation/testbench-known-issues.asciidoc#L75)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

`-Dcom.vaadin.testbench.Parameters.maxAttempts=2` is the primary way. I cannot see why somebody would want to use `System.setProperty` for this, especially if you add `Parameters.setMaxAttempts(2)`.
</blockquote></details>

As I mentioned before, probably you want to use this Parameters.maxAttempts at all, because it will affect all your tests.
What you want to do is to pick one/two randomly failing tests and add a RetryRule over there.
I've just mentioned this Parameters thing, because it's used for framework. 

---

*[documentation/testbench-known-issues.asciidoc, line 78 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxU1xYBH8WY0CLHu8b:-KYxd8bgRVP4HpTNBpNl:b-896fix) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/documentation/testbench-known-issues.asciidoc#L78)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

remove ","
</blockquote></details>

Done.

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/Parameters.java, line 426 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxTswP7qKU-NaMGxCI:-KYxfWPTKr5obBRXjHub:b-896fix) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/vaadin-testbench-core/src/main/java/com/vaadin/testbench/Parameters.java#L426)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

All other parameters also have a setter
</blockquote></details>

Done.

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/Parameters.java, line 427 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxTc6nh-92sD0MlLUw:-KYxfTKs57XUNxXdn2cy:b-896fix) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/vaadin-testbench-core/src/main/java/com/vaadin/testbench/Parameters.java#L427)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

this should be maxAttempts to be consistent with all other parameters
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/testbench/852)*
<!-- Sent from Reviewable.io -->

Artur- commented:
Reviewed 3 of 4 files at r8.
Review status: 9 of 10 files reviewed at latest revision, 5 unresolved discussions.

---

*[documentation/testbench-known-issues.asciidoc, line 75 at r7](https://reviewable.io:443/reviews/vaadin/testbench/852#-KYxTHlGH7lUtVXxPV74:-KZ0ln5wp7BqBK4Xjhvz:bq449ig) ([raw file](https://github.com/vaadin/testbench/blob/8fade62290edafa90ea93a650909227bf946c77c/documentation/testbench-known-issues.asciidoc#L75)):*
<details><summary><i>Previously, rogozinds (rogozinds) wrote…</i></summary><blockquote>

As I mentioned before, probably you want to use this Parameters.maxAttempts at all, because it will affect all your tests.
What you want to do is to pick one/two randomly failing tests and add a RetryRule over there.
I've just mentioned this Parameters thing, because it's used for framework. 
</blockquote></details>

Still should use `-Dcom.vaadin.testbench.Parameters.maxAttempts=2` instead of `System.setProperty`

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/TestBenchTestCase.java, line 43 at r8](https://reviewable.io:443/reviews/vaadin/testbench/852#-KZ0mv-WoGzWElh9e_HS:-KZ0mv-WoGzWElh9e_HT:b-dm1sta) ([raw file](https://github.com/vaadin/testbench/blob/86a3b3c7846da0a3a8726ad4bf40aa633604ee31/vaadin-testbench-core/src/main/java/com/vaadin/testbench/TestBenchTestCase.java#L43)):*
> ```Java
>     /**
>      * Specifies retry count, which is used to run same test several times.
>      * Can be changed by setting "maxAttempts" system property.
> ```

Change to fully qualified property name

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 1 at r8](https://reviewable.io:443/reviews/vaadin/testbench/852#-KZ0mc1dX6soe7v_wXrB:-KZ0mc1ey_lD2BT4W6wx:b-v8gwnn) ([raw file](https://github.com/vaadin/testbench/blob/86a3b3c7846da0a3a8726ad4bf40aa633604ee31/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L1)):*
> ```Java
> package com.vaadin.testbench.annotations;
> ```

Seems to be in the wrong package

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/testbench/852)*
<!-- Sent from Reviewable.io -->

Artur- commented:
Reviewed 1 of 4 files at r8.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/testbench/852)*
<!-- Sent from Reviewable.io -->

Artur- commented:
Reviewed 16 of 16 files at r9.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/testbench/852)*
<!-- Sent from Reviewable.io -->

rogozinds commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/TestBenchTestCase.java, line 43 at r8](https://reviewable.io:443/reviews/vaadin/testbench/852#-KZ0mv-WoGzWElh9e_HS:-KZ0oBhZnkSYEOM-7Lgz:b-896fix) ([raw file](https://github.com/vaadin/testbench/blob/86a3b3c7846da0a3a8726ad4bf40aa633604ee31/vaadin-testbench-core/src/main/java/com/vaadin/testbench/TestBenchTestCase.java#L43)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

Change to fully qualified property name
</blockquote></details>

Done.

---

*[vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java, line 1 at r8](https://reviewable.io:443/reviews/vaadin/testbench/852#-KZ0mc1dX6soe7v_wXrB:-KZ0oD6vgK9Qlw29NQqK:b-896fix) ([raw file](https://github.com/vaadin/testbench/blob/86a3b3c7846da0a3a8726ad4bf40aa633604ee31/vaadin-testbench-core/src/main/java/com/vaadin/testbench/annotations/RetryRule.java#L1)):*
<details><summary><i>Previously, Artur- (Artur) wrote…</i></summary><blockquote>

Seems to be in the wrong package
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/testbench/852)*
<!-- Sent from Reviewable.io -->
